### PR TITLE
Fix #924: Empty data string is not a file name

### LINKED ISSUE #924: Calling render without file or data produces a cryptic error message
On master, and doing `sass.render({contents: input}, callback)`.

When `input` is an empty string, I get the error `File context created without an input path`.

### PR DESCRIPTION
**Pull request by saper**

### CODE REVIEW

File: test/api.js [hunk 1]
```diff
@@ -76,6 +76,15 @@ describe('api', function() {
       });
     });
 
+    it('should NOT compile empty data string', function(done) {
+      sass.render({
+        data: ''
+      }, function(error) {
+        assert.equal(error.message, 'Data context created with empty source string');
```
Comment (xzyfer) [inline test/api.js]: IMO this error message is the real problem. We need to validate better in the CLI and return and error message that helps the user.

Comment (saper) [inline test/api.js]: I am kind of ambivalent about this:

- we are pretty thin wrapper here and I wouldn't really introduce own `node-sass` error messages if `libsass` already handles the situation
- I am not sure that having two kinds of contexts (data and file) are good design. They are actually the same structures differing just by one extra field may be very error prone. A data context with `source_string` set to zero is indistinguishable from the file context - we have to use two sets of functions to handle the distinction. We also have ambiguous situation when both `file` and `data` are set - there is a special meaning for this situation as well (see #914).

We need some work on the libsass side to have this ironed out in my opinion.

Comment (xzyfer) [inline test/api.js]: > we are pretty thin wrapper here

This is no longer the case in 3.x now that we have custom functions and types. We need to stop thinking this way.

>I wouldn't really introduce own node-sass error messages if libsass already handles the situation

We should serve our users as best we can. If Libsass creates a poor user experience we should fill those gaps until they're resolved upstream. It's important to understand that Libsass has different goals to us for the time being.

> I am not sure that having two kinds of contexts (data and file) is a good design

I don't disagree here. I think we should still take the two inputs (maybe more in the future?) but having two separate properties for this doesn't serve a benefit to the user IMHO.

>We need some work on the libsass side to have this ironed out in my opinion.

I agree that Libsass needs to provide a better developer experience.

Long story short I emphatically disagree that this is sane behaviour.

Any message we show to the user should clearly describe what went wrong and how to resolve it. IMHO `Data context created with empty source string` does neither.

**Comment by saper** [comment]
`renderSync({})` hangs - to be investigated :) volunteers needed!

**Comment by saper** [comment]
Once tests pass I think this is ready for review. I was afraid there was another binding bug hiding, but fortunately it was just a missing `done()` in the mocha script. @xzyfer

File: lib/index.js [hunk 2]
```diff
@@ -351,7 +352,13 @@ module.exports.render = function(options, cb) {
     });
   }
 
-  return options.data ? binding.render(options) : binding.renderFile(options);
+  if (options.hasOwnProperty('data')) {
+    binding.render(options);
+  } else if (options.hasOwnProperty('file')) {
+    binding.renderFile(options);
+  } else {
+    cb({status: 3, message: 'No input specified: provide a file name or a source string to process' });
```
Comment (xzyfer) [inline lib/index.js]: :heart_eyes:

**Comment by xzyfer** [comment]
:+1:

File: lib/index.js [hunk 3]
```diff
@@ -351,7 +352,13 @@ module.exports.render = function(options, cb) {
     });
   }
 
-  return options.data ? binding.render(options) : binding.renderFile(options);
+  if (options.hasOwnProperty('data')) {
```
Comment (xzyfer) [inline lib/index.js]: IMHO these checks are too lax. 

```
if (options.hasOwnProperty('data') && options.data) {
  binding.render(options);
} else if (options.hasOwnProperty('file') && options.file) {
  binding.renderFile(options);
} else {
 ...
}
```

A use should _never_ see

> Data context created with empty source string

File: lib/index.js [hunk 4]
```diff
@@ -398,7 +405,15 @@ module.exports.renderSync = function(options) {
     });
   }
 
-  var status = options.data ? binding.renderSync(options) : binding.renderFileSync(options);
+  var status;
+  if (options.hasOwnProperty('data')) {
```
Comment (xzyfer) [inline lib/index.js]: ditto

**Comment by xzyfer** [comment]
The new error message is big improvement. However I'm still argue that under no circumstance should a user receive this error message if we can be prevented (which in this case it can)

>Data context created with empty source string

**Comment by saper** [comment]
In that case, I think `if (options.data) { ... ` and `if (options.file) { ... ` are enough?

**Comment by xzyfer** [comment]
Agreed

**Comment by saper** [comment]
in that case the last commit should be ok
